Broken on 1.4.0

Still investigating to figure out what happened but all usage of `$path(...)` is broken in 1.4.0 but works again when I revert to 1.3.1.


## Sample Error
```txt
app/routes/org/$orgId/app/index.tsx:351:23 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '"/org/:orgId/app/:appId/:branch/:componentType"' is not assignable to parameter of type '"/org"'.

351             to={$path("/org/:orgId/app/:appId/:branch/:componentType", {
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/remix-routes.d.ts:521:25
    521 export declare function $path(
                                ~~~~~
    The last overload is declared here.

app/routes/org/$orgId/index.tsx:8:25 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '"/org/:orgId/app"' is not assignable to parameter of type '"/org"'.

8   return redirect($path("/org/:orgId/app", { orgId: orgId as string }));
                          ~~~~~~~~~~~~~~~~~

  node_modules/remix-routes.d.ts:521:25
    521 export declare function $path(
                                ~~~~~
    The last overload is declared here.

```

## Notes
I'm also using Typescript "strict" mode

--- a/e2e/app/entry.client.ts
+++ b/e2e/app/entry.client.ts
@@ -1,0 +1,1 @@
+export default function() {}

--- a/e2e/app/entry.server.ts
+++ b/e2e/app/entry.server.ts
@@ -1,0 +1,1 @@
+export default function() {}

--- a/e2e/app/root.tsx
+++ b/e2e/app/root.tsx
@@ -1,0 +1,3 @@
+export default function Root() {
+  return <div></div>;
+}

--- a/e2e/app/routes/($lang)/about.tsx
+++ b/e2e/app/routes/($lang)/about.tsx
@@ -1,0 +1,8 @@
+export default function Component() {
+  return (
+    <div>
+      <h1>About</h1>
+      <p>This is the about page.</p>
+    </div>
+  );
+}

--- a/e2e/app/routes/index.tsx
+++ b/e2e/app/routes/index.tsx
@@ -1,21 +1,7 @@
-import { $params, $path } from 'remix-routes';
-
-import { Routes } from 'remix-routes';
-import { redirect as remixRedirect } from "@remix-run/node";
-
-function redirect<
-  R extends keyof Routes,
-  P extends Routes[R]["params"],
-  Q extends Routes[R]["query"]
->(route: R, params: P, query?: Q) {
-  const includeParams = route.indexOf('/:') > -1;
-  return remixRedirect(
-      includeParams ? $path(route as any, params as any, query as any) : $path(route as any, query as any)
-  );
+export default function Component() {
+  return (
+    <div>
+      <h1>Index</h1>
+    </div>
+  )
 }
-
-redirect("/posts/:id", { id: 1 }, { view: "list", sort: "price" });
-
-$path('/posts/:id', { id: 1 }, { view: 'list', sort: 'price' });
-
-$path('/', { foo: 'bar' });

--- a/e2e/app/routes/posts/$id.tsx
+++ b/e2e/app/routes/posts/$id.tsx
@@ -1,4 +1,5 @@
 import { ActionFunction, LoaderFunction } from '@remix-run/node';
+import { useCatch } from '@remix-run/react';
 
 export type SearchParams = {
   view: 'list' | 'grid';

--- a/e2e/package.json
+++ b/e2e/package.json
@@ -9,14 +9,14 @@
     "@remix-run/node": "^1.9.0",
     "@remix-run/react": "^1.9.0",
     "react": "^18.0.0",
-    "remix-routes": "workspace:*",
-    "typescript": "^4.4.3"
+    "remix-routes": "workspace:*"
   },
   "devDependencies": {
     "@remix-run/dev": "^1.9.0",
     "@types/react": "^18.0.9",
     "readline": "^1.3.0",
-    "strip-indent": "^4.0.0"
+    "strip-indent": "^4.0.0",
+    "typescript": "^4.5.4"
   },
   "version": null
 }

--- a/e2e/tests/types.test.ts
+++ b/e2e/tests/types.test.ts
@@ -1,0 +1,44 @@
+import { expectTypeOf, test } from 'vitest';
+import { $params, $path, Routes } from "remix-routes";
+
+type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
+
+test('$path', () => {
+  expectTypeOf($path('/')).toEqualTypeOf<string>();
+  expectTypeOf($path('/', { foo: "bar" })).toEqualTypeOf<string>();
+  expectTypeOf($path('/posts')).toEqualTypeOf<string>();
+  expectTypeOf($path('/posts', { foo: "bar" })).toEqualTypeOf<string>();
+  expectTypeOf($path('/posts/:id', { id: 1 }, { view: "list", sort: 'price' })).toEqualTypeOf<string>();
+  expectTypeOf($path('/:lang?/about', { lang: "en" })).toEqualTypeOf<string>();
+});
+
+test('$prams', () => {
+  let params: any = {};
+  expectTypeOf($params('/posts/:id', params)).toEqualTypeOf<{ id: string }>();
+  expectTypeOf($params('/:lang?/about', params)).toEqualTypeOf<{ lang?: string }>();
+});
+
+test('Routes', () => {
+  expectTypeOf<Routes['/']>().toMatchTypeOf<{
+    params: {},
+    query: URLSearchParamsInit,
+  }>();
+
+  expectTypeOf<Routes['/:lang?/about']>().toMatchTypeOf<{
+    params: { lang?: string | number },
+    query: URLSearchParamsInit,
+  }>();
+
+  expectTypeOf<Routes['/posts']>().toMatchTypeOf<{
+    params: {},
+    query: URLSearchParamsInit,
+  }>();
+
+  expectTypeOf<Routes['/posts/:id']>().toEqualTypeOf<{
+    params: { id: string | number },
+    query: {
+      view: 'list' | 'grid';
+      sort?: 'price' | 'size';
+    },
+  }>();
+});

--- a/package.json
+++ b/package.json
@@ -16,6 +16,6 @@
   "devDependencies": {
     "@changesets/cli": "^2.26.0",
     "typescript-remix-routes-plugin": "workspace:*",
-    "vitest": "^0.23.2"
+    "vitest": "^0.28.5"
   }
 }

--- a/packages/remix-routes/src/cli.ts
+++ b/packages/remix-routes/src/cli.ts
@@ -111,7 +111,8 @@ type IsAny<T> = (
     : false
 );
 type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
-type Query<T> = IsAny<T> extends true ? URLSearchParamsInit : T;
+type ExportedQuery<T> = IsAny<T> extends true ? URLSearchParamsInit : T;
+type Query<T> = IsAny<T> extends true ? [URLSearchParamsInit?] : [T];
       `,
       generateRouteDefinition(routesInfo),
       generatePathDefinition(routesInfo),
@@ -138,7 +139,7 @@ function generateRouteDefinition(routesInfo: RoutesInfo) {
       (param) => `${param}: string | number`,
     );
     lines.push(`    params: { ${paramsType.join('; ')} },`);
-    lines.push(`    query: Query<import('../app/${slash(fileName.replace(/\.tsx?$/, ''))}').SearchParams>,`)
+    lines.push(`    query: ExportedQuery<import('../app/${slash(fileName.replace(/\.tsx?$/, ''))}').SearchParams>,`)
     lines.push('  };')
     code.push(lines.join('\n'));
   });
@@ -155,7 +156,7 @@ function generatePathDefinition(routesInfo: RoutesInfo) {
       lines.push(`  params: Routes["${route}"]["params"],`);
     }
     lines.push(
-      `  ...query: [Routes["${route}"]["query"]]`,
+      `  ...query: Query<import('../app/${slash(fileName.replace(/\.tsx?$/, ''))}').SearchParams>`,
     );
     lines.push(`): string;`);
     code.push(lines.join('\n'));

--- a/vitest.config.ts
+++ b/vitest.config.ts
@@ -3,6 +3,6 @@ import { defineConfig } from 'vitest/config';
 export default defineConfig({
   test: {
     testTimeout: 10_000,
-    exclude: ['**/*.js', 'node_modules']
+    exclude: ['**/*.js', 'node_modules', 'e2e/node_modules']
   },
 });

